File: rl-trainer-editor/backend/planner/observation_catalog.py
"""Observation catalog — procedural groups + sensor instances from the project."""
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any, Literal

from domain.models import ObservationTerm
from planner.observation_normalization import apply_recommended_normalization
from storage import project_storage

ObservationSource = Literal["procedural", "sensor"]


@dataclass(frozen=True)
class ProceduralCatalogEntry:
    id: str
    label: str
    kind: str
    category: str
    description: str


PROCEDURAL_CATALOG: tuple[ProceduralCatalogEntry, ...] = (
    ProceduralCatalogEntry(
        "joint_positions",
        "Joint positions",
        "joint_state",
        "state",
        "Relative joint positions from the sim / ros2_control state.",
    ),
    ProceduralCatalogEntry(
        "joint_velocities",
        "Joint velocities",
        "joint_state",
        "state",
        "Joint velocities from the sim / ros2_control state.",
    ),
    ProceduralCatalogEntry(
        "last_actions",
        "Last actions",
        "action",
        "state",
        "Previous policy action vector (action history for smoothness).",
    ),
    ProceduralCatalogEntry(
        "commands",
        "Commands",
        "command",
        "command",
        "Target velocity, height, and gait command from the task.",
    ),
    ProceduralCatalogEntry(
        "base_lin_vel",
        "Base linear velocity",
        "base_state",
        "state",
        "Base linear velocity in body frame (odom or kinematics).",
    ),
    ProceduralCatalogEntry(
        "base_ang_vel",
        "Base angular velocity",
        "base_state",
        "state",
        "Base angular velocity (IMU or kinematics).",
    ),
    ProceduralCatalogEntry(
        "projected_gravity",
        "Projected gravity",
        "orientation",
        "state",
        "Gravity vector in body frame (IMU orientation or equivalent).",
    ),
)

_CATEGORY_SENSOR_KINDS: dict[str, set[str]] = {
    "contact": {"contact"},
    "orientation": {"imu"},
    "posture": {"imu"},
    "gait": {"contact"},
}


def _slug(text: str) -> str:
    return re.sub(r"[^a-z0-9_]+", "_", text.lower()).strip("_")


def _observation_key(name: str) -> str:
    return name.replace("-", "_").replace(" ", "_")


def _fields_for_sensor(kind: str, sensor: dict[str, Any]) -> list[str]:
    if kind == "imu":
        fields = ["angular_velocity", "linear_acceleration"]
        imu = sensor.get("imu") or {}
        if imu.get("enableOrientation"):
            fields.append("orientation")
        return fields
    if kind == "contact":
        return ["contacts"]
    if kind == "odom":
        odom = sensor.get("odom") or {}
        dims = odom.get("dimensions", 3)
        if dims == 2:
            return ["linear_velocity_x", "angular_velocity_z"]
        return ["linear_velocity_x", "linear_velocity_y", "angular_velocity_z"]
    if kind == "lidar":
        return ["ranges", "angle_min", "angle_max"]
    return []


def _load_sensor_model(name: str) -> dict[str, Any] | None:
    path = project_storage.sensor_model_path(name)
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text())
        return data if isinstance(data, dict) else None
    except (json.JSONDecodeError, OSError):
        return None


def _sensor_terms_from_model(project: str) -> list[ObservationTerm]:
    doc = _load_sensor_model(project)
    if not doc:
        return _sensor_terms_from_yaml(project)

    sensors = doc.get("sensors") or []
    if not isinstance(sensors, list):
        return []

    terms: list[ObservationTerm] = []
    for sensor in sensors:
        if not isinstance(sensor, dict):
            continue
        sensor_id = str(sensor.get("id") or "")
        kind = str(sensor.get("kind") or "").lower()
        name = str(sensor.get("name") or sensor_id or kind)
        key = _observation_key(name)
        terms.append(
            ObservationTerm(
                id=f"sensor:{sensor_id or key}",
                source="sensor",
                kind=kind,
                category="sensor",
                label=name,
                enabled=False,
                available=bool(sensor.get("enabled", True)),
                key=key,
                topic=str(sensor.get("rosTopic") or ""),
                parentLink=str(sensor.get("parentLink") or ""),
                rateHz=float(sensor.get("updateRate") or 0),
                fields=_fields_for_sensor(kind, sensor),
                sensorId=sensor_id,
            )
        )
    return terms


def _sensor_terms_from_yaml(project: str) -> list[ObservationTerm]:
    doc = project_storage.load_observations_doc(project)
    if not doc:
        return []
    obs = doc.get("observations") or {}
    if not isinstance(obs, dict):
        return []

    terms: list[ObservationTerm] = []
    for key, spec in obs.items():
        if not isinstance(spec, dict):
            continue
        kind = str(spec.get("kind") or "").lower()
        terms.append(
            ObservationTerm(
                id=f"sensor:{key}",
                source="sensor",
                kind=kind,
                category="sensor",
                label=key,
                enabled=False,
                available=True,
                key=key,
                topic=str(spec.get("topic") or ""),
                msgType=str(spec.get("msg_type") or ""),
                parentLink=str(spec.get("parent_link") or ""),
                rateHz=float(spec.get("rate_hz") or 0),
                fields=list(spec.get("fields") or []),
                sensorId=key,
            )
        )
    return terms


def build_observation_catalog(project: str) -> list[ObservationTerm]:
    """Full catalog: procedural entries + all sensor instances for the project."""
    terms: list[ObservationTerm] = []
    for entry in PROCEDURAL_CATALOG:
        terms.append(
            ObservationTerm(
                id=entry.id,
                source="procedural",
                kind=entry.kind,
                category=entry.category,
                label=entry.label,
                enabled=False,
                available=True,
                key=entry.id,
                description=entry.description,
            )
        )
    terms = [apply_recommended_normalization(t) for t in terms]
    terms.extend(_sensor_terms_from_model(project))
    return [apply_recommended_normalization(t) for t in terms]


def merge_observation_terms(
    existing: list[ObservationTerm] | None,
    project: str,
) -> list[ObservationTerm]:
    """Merge saved selection with the latest project catalog."""
    catalog = build_observation_catalog(project)
    by_id = {t.id: t.model_copy(deep=True) for t in (existing or [])}
    merged: list[ObservationTerm] = []
    for cat in catalog:
        prev = by_id.get(cat.id)
        if prev:
            term = cat.model_copy(deep=True)
            term.enabled = prev.enabled if prev.available or cat.available else False
            term.scale = prev.scale if prev.scale > 0 else cat.scale
            term.offset = prev.offset
            term.clipMin = prev.clipMin
            term.clipMax = prev.clipMax
            if prev.label and cat.source == "sensor":
                term.label = prev.label
            merged.append(term)
        else:
            merged.append(cat)
    return merged


def recommend_observation_terms(
    terms: list[ObservationTerm],
    *,
    reward_categories: set[str],
) -> tuple[list[ObservationTerm], list[str]]:
    """Enable a recommended subset based on enabled reward categories."""
    needs_contact = bool({"contact", "gait"} & reward_categories)
    needs_imu = bool({"orientation", "posture"} & reward_categories)
    needs_velocity = bool({"velocity", "tracking"} & reward_categories)
    needs_action = bool({"action_smoothness"} & reward_categories)

    core_procedural = {
        "joint_positions",
        "joint_velocities",
        "last_actions",
        "commands",
    }
    velocity_procedural = {"base_lin_vel", "base_ang_vel"}
    orientation_procedural = {"projected_gravity"}

    out: list[ObservationTerm] = []
    notes: list[str] = []

    for term in terms:
        t = term.model_copy(deep=True)
        if not t.available:
            t.enabled = False
            out.append(t)
            continue

        if t.source == "procedural":
            enable = t.id in core_procedural
            if needs_velocity and t.id in velocity_procedural:
                enable = True
            if needs_imu and t.id in orientation_procedural:
                enable = True
            if needs_action and t.id == "last_actions":
                enable = True
            t.enabled = enable
        else:
            kind = t.kind.lower()
            enable = False
            if kind == "contact" and needs_contact:
                enable = True
            elif kind == "imu" and needs_imu:
                enable = True
            elif kind == "odom" and needs_velocity:
                enable = True
            elif kind == "lidar":
                enable = False
            t.enabled = enable
        out.append(t)

    enabled_count = sum(1 for t in out if t.enabled)
    out = [apply_recommended_normalization(t) for t in out]
    notes.append(
        f"Recommended {enabled_count} of {len(out)} observations "
        f"for reward categories: {sorted(reward_categories) or ['default']}."
    )
    return out, notes

File: packages/rl-trainer-model/observationCatalog.ts
/** Procedural observation groups assembled in the RL env (not separate ROS sensors). */

export type ObservationSource = "procedural" | "sensor";

export interface ObservationNormDefaults {
  scale: number;
  offset: number;
  clipMin: number | null;
  clipMax: number | null;
}

export interface ObservationCatalogEntry {
  id: string;
  label: string;
  kind: string;
  category: "state" | "command" | "sensor";
  description: string;
  dimsHint?: string;
}

export const PROCEDURAL_OBSERVATION_CATALOG: ObservationCatalogEntry[] = [
  {
    id: "joint_positions",
    label: "Joint positions",
    kind: "joint_state",
    category: "state",
    description: "Relative joint positions from the sim / ros2_control state.",
    dimsHint: "n_joints",
  },
  {
    id: "joint_velocities",
    label: "Joint velocities",
    kind: "joint_state",
    category: "state",
    description: "Joint velocities from the sim / ros2_control state.",
    dimsHint: "n_joints",
  },
  {
    id: "last_actions",
    label: "Last actions",
    kind: "action",
    category: "state",
    description: "Previous policy action vector (action history for smoothness).",
    dimsHint: "n_actions",
  },
  {
    id: "commands",
    label: "Commands",
    kind: "command",
    category: "command",
    description: "Target velocity, height, and gait command from the task.",
    dimsHint: "4–5",
  },
  {
    id: "base_lin_vel",
    label: "Base linear velocity",
    kind: "base_state",
    category: "state",
    description: "Base linear velocity in body frame (odom or kinematics).",
    dimsHint: "3",
  },
  {
    id: "base_ang_vel",
    label: "Base angular velocity",
    kind: "base_state",
    category: "state",
    description: "Base angular velocity (IMU or kinematics).",
    dimsHint: "3",
  },
  {
    id: "projected_gravity",
    label: "Projected gravity",
    kind: "orientation",
    category: "state",
    description: "Gravity vector in body frame (IMU orientation or equivalent).",
    dimsHint: "3",
  },
];

export const OBSERVATION_CATEGORY_HINTS: Record<string, string> = {
  state: "Procedural state from sim / controller feedback",
  command: "Task command reference included in the observation vector",
  sensor: "ROS topic bridged from Gazebo (see sens_*_observations.yaml)",
};

export const OBSERVATION_KIND_HINTS: Record<string, string> = {
  joint_state: "Requires ros2_control joint state interface",
  action: "Uses previous policy output from the env wrapper",
  command: "Uses stage command targets from RL config",
  base_state: "Uses odometry or differentiated base pose",
  orientation: "Uses IMU orientation or projected gravity",
  imu: "sensor_msgs/Imu — angular velocity, acceleration, optional orientation",
  contact: "Foot contact / slip from Gazebo contact sensors",
  odom: "Base velocity from odometry source",
  lidar: "Range scan (usually disabled for locomotion PPO)",
};

const catalogById = new Map(PROCEDURAL_OBSERVATION_CATALOG.map((e) => [e.id, e]));

export const PROCEDURAL_NORM: Record<string, ObservationNormDefaults> = {
  joint_positions: { scale: 2, offset: 0, clipMin: -1, clipMax: 1 },
  joint_velocities: { scale: 6, offset: 0, clipMin: -1, clipMax: 1 },
  last_actions: { scale: 1, offset: 0, clipMin: -1, clipMax: 1 },
  commands: { scale: 1, offset: 0, clipMin: -1, clipMax: 1 },
  base_lin_vel: { scale: 2, offset: 0, clipMin: -1, clipMax: 1 },
  base_ang_vel: { scale: 8, offset: 0, clipMin: -1, clipMax: 1 },
  projected_gravity: { scale: 1, offset: 0, clipMin: null, clipMax: null },
};

export const SENSOR_KIND_NORM: Record<string, ObservationNormDefaults> = {
  imu: { scale: 5, offset: 0, clipMin: -1, clipMax: 1 },
  contact: { scale: 1, offset: 0, clipMin: 0, clipMax: 1 },
  odom: { scale: 2, offset: 0, clipMin: -1, clipMax: 1 },
  lidar: { scale: 30, offset: 0, clipMin: 0, clipMax: 1 },
};

export const OBSERVATION_NORM_FORMULA =
  "Env applies: clip((raw − offset) / scale, clipMin, clipMax). Omit clip when min/max are empty.";

export const OBSERVATION_NORM_HINTS: Record<string, string> = {
  scale: "Divisor — typical magnitude of raw values (e.g. 2 for joint rad, 6 for rad/s).",
  offset: "Subtract from raw value before scaling (usually 0).",
  clipMin: "Lower clip after scaling; leave empty for no lower bound.",
  clipMax: "Upper clip after scaling; leave empty for no upper bound.",
};

export function getRecommendedObservationNorm(term: {
  id: string;
  source: ObservationSource;
  kind: string;
}): ObservationNormDefaults {
  if (term.source === "procedural" && term.id in PROCEDURAL_NORM) {
    return { ...PROCEDURAL_NORM[term.id] };
  }
  const kind = term.kind.toLowerCase();
  if (kind in SENSOR_KIND_NORM) {
    return { ...SENSOR_KIND_NORM[kind] };
  }
  return { scale: 1, offset: 0, clipMin: -1, clipMax: 1 };
}

export function getProceduralObservationEntry(id: string): ObservationCatalogEntry | undefined {
  return catalogById.get(id);
}
